### ISSUE TITLE
Confusing "parameter flow" for the calculators

### ISSUE THREAD

**Issue by ceriottm** [lab-cosmo/torch-pme#57]
The calculators use a mix of class parameters and "compute function" parameters to define e.g. the smearing and the calculation accuracy. This is confusing and a bit clunky - it'd be better to streamline the implementation to avoid having to worry too much about (re)setting these values.

**Comment by PicoCentauri** [comment]
We had some brainstorming meeting with @E-Rum and @GardevoirX and came up some refined structure for the calculator classes.

Basically, we will split the calculators into classes of EwaldCalculators (Ewald, PME, P3M) and a single DirectCalculators. This means we don't have to branch a lot if a potential has a finite `smearing` or not. The proposed structure should look like

```python
import torch


class CalcultorBase(torch.nn.Module):
    def __init__(self): ...

    def forward(self):
        # validate the input and call the _compute method
        return self._compute()

    def _compute(self):
        raise NotImplementedError("Only implemented in subclasses")


class EwaldCalculator(CalcultorBase):
    def __init__(self, potential):
        super().__init__()

        if potential.smearing is None:
            raise ValueError()

    def _compute(self):
        return (
            self._compute_kspace()
            + self._compute_rspace()
            + self._compute_background_contribution()
            + self._compute_self_contriubution()
        )

    def _compute_background_contribution(self): ...

    def _compute_self_contriubution(self): ...

    def _compute_kspace(self): ...

    def _compute_rspace(self): ...


class PMECalculator(EwaldCalculator):
    def __init__(self): ...

    def _compute_kspace(self):
        # override `_compute_kspace` from Ewald with PME logic
        ...


class P3MCalculator(EwaldCalculator):
    def __init__(self): ...

    def _compute_kspace(self): ...


class DirectCalculator(CalcultorBase):
    def __init__(self, potential):

        if potential.smearing is not None:
            raise ValueError()

    def _compute(self):
        # call potential.from_dist() and multiply the charges
        ...
```

As you see this will not break API at all and should simplify the parameter flow. If there are no objection, @E-Rum volunteered to check the implementation and do it. THANKS A LOT ALREADY!!!!
